### ISSUE TITLE
DataFetcherResult support to return data & errors

### ISSUE THREAD

**Issue by elenigen** [leangen/graphql-spqr#78]
DataFetcherResult was introduced in the graphql-java v7.0 to allow returning a wrapper object which will contain the data the resolver is supposed to return + a list of errors.

I took the sample project to try:
```java
    @GraphQLMutation(name = "createVendor")
    public DataFetcherResult<Vendor> createVendor(@GraphQLArgument(name = "vendor") Vendor vendor){
        Vendor createdVendor =  new Vendor((long) mockVendorStorage.size() + 1,
                vendor.getName(),
                vendor.getAddress());
        mockVendorStorage.add(createdVendor);
        List<GraphQLError> errors = new ArrayList<>();
        errors.add(new GenericGraphQLError("invalid data"));
        return new DataFetcherResult<>(createdVendor, errors);
    }
```
Then I invoke the method:
```js
mutation {
  createVendor(
    vendor: {
      id: 1
      name: "hello"
      address: {
        city: "mtl"
        postalCode:"h2c1f1"
        streetAndNumber: "123 asd"
      }
      productsInStock:[
        {
          stockSize: 1
          product:{
            id: 1
            name: "aaa"
            description:"asdas"
          }
        }
      ]
    }) {
    id
    name
  }
}
```

Instead of returning the data + the error, it's just failing because spqr hasn't extracted the data part from the DataFetcherResult:
```js
{
  "data": null,
  "errors": [
    {
      "message": "Validation error of type FieldUndefined: Field 'id' in type 'DataFetcherResult_Vendor' is undefined",
      "locations": [
        {
          "line": 36,
          "column": 5
        }
      ],
      "description": "Field 'id' in type 'DataFetcherResult_Vendor' is undefined",
      "validationErrorType": "FieldUndefined",
      "errorType": "ValidationError",
      "path": null,
      "extensions": null
    },
    {
      "message": "Validation error of type FieldUndefined: Field 'name' in type 'DataFetcherResult_Vendor' is undefined",
      "locations": [
        {
          "line": 37,
          "column": 5
        }
      ],
      "description": "Field 'name' in type 'DataFetcherResult_Vendor' is undefined",
      "validationErrorType": "FieldUndefined",
      "errorType": "ValidationError",
      "path": null,
      "extensions": null
    }
  ],
  "extensions": null
}
```

Are you planning to support this new feature from graphql-java implementation, since actually it was a big missing peace from the GraphQL spec.

**Comment by kaqqao** [comment]
Actually the mapper handling this is [already present](https://github.com/leangen/graphql-spqr/blob/master/src/main/java/io/leangen/graphql/generator/mapping/core/DataFetcherResultMapper.java) in the current `master`. So you'll have it in the next release. But, I'd like to make a few points:

1. It's good to notice the mapper is almost trivial to add yourself. See its code, it's only a couple of lines. Once you have a custom mapper, you can register it via:

```java
generator.withTypeMappers((config, defaults) -> defaults.insertAfter(StreamToCollectionTypeAdapter.class, customMapper) //for example, other locations are also good
```

2. I really can't recommend using `DataFetcherResult` with SPQR in the first place. SPQR is all about non-obtrusiveness and using a library-dependent type in your code goes directly against this. You end up having a service that can not be (naturally) reused outside of GraphQL due to its tight coupling with graphql-java.
This is especially aggravating since you don't need `DataFetcherResult` to achieve the same benefits, the facilities for a much cleaner approach are already there: the method can return its real result or throw a custom exception that contains one or more messages. You then provide a `DataFetcherExceptionHandler` to the `ExecutionStrategy` that will map the custom exception to one or more errors in the response (using `ExecutionContext#addError`) e.g one error per message.

This way your Java code remains succinct and reusable. You could, for example, use the exact same method as a backend for a REST end-point. Do you agree?

**Comment by kaqqao** [comment]
Oh, wait. The mapper is [already present](https://github.com/leangen/graphql-spqr/blob/graphql-spqr-v0.9.6/src/main/java/io/leangen/graphql/generator/mapping/core/DataFetcherResultMapper.java) in the released version 0.9.6. So no need to wait for anything. Were you on an older release?

**Comment by elenigen** [comment]
@kaqqao Thanks for the reply, actually I simply did a checkout of the sample project, but I haven't noticed that it's using an older version.  I tried with `0.9.6` as you suggested, the result is better, but still not what I expected.  So far in our project, that's the biggest blocker, because for some reason (or we misunderstood something) all the libs using graphql-java are not following the [GraphQL specification:](http://facebook.github.io/graphql/October2016/#sec-Errors)
> If the data entry in the response is not null, the errors entry in the response may contain any errors that occurred during execution. If errors occurred during execution, it should contain those errors.

So that's exactly what we are trying to get data + errors in the same response.  Just to be clear, for example, the request I added earlier is supposed to get this as response:
```js
{
  "data": {
      "id": 1,
      name: "hello"
   },
  "errors": [
    {
      "locations": null,
      "message": "invalid data",
      "errorType": null,
      "path": null,
      "extensions": null
    }
  ],
  "extensions": null
}
```

With the version `0.9.6`, I only get the error:
```js
{
  "data": null,
  "errors": [
    {
      "locations": null,
      "message": "invalid data",
      "errorType": null,
      "path": null,
      "extensions": null
    }
  ],
  "extensions": null
}
```

In other words, I totally agree with you we should avoid using `DataFetcherResult` as much as we can, but in some specific scenario, it's just not possible, because how do you want to return data and throw an exception at the same time? ... that's the whole point of introducing the `DataFetcherResult', a result can be partially good to return some data and partially bad because of an error.

So, I would believe this issue should still be open, what do you think?

**Comment by kaqqao** [comment]
As for the spec, it does allow both the data and the errors. But this usually means that you get the data for the successfully resolved nodes and errors for unsuccessful ones. I find it very strange for the same node to produce both, especially in Java, but it is allowed.
I bumped the version in the sample to 0.9.6, thanks for pointing that out!
As for the issue you're describing, I'm still not sure I'm getting it... Here's my quick attempt at replicating it:

```java
public static class Service {

        @GraphQLQuery
        public DataFetcherResult<String> mixed() {

            return new DataFetcherResult<>("result", Collections.singletonList(new GraphQLError() {
                @Override
                public String getMessage() {
                    return "Something went wrong";
                }

                @Override
                public List<SourceLocation> getLocations() {
                    return null;
                }

                @Override
                public ErrorType getErrorType() {
                    return ErrorType.ValidationError;
                }
            }));
        }
    }

...

ExecutionResult result = graphQl.execute("{mixed}");
```
After this, `result.data.mixed` contains the string `result`, while `result.errors[0].message` contains `Something went wrong`, as I'd expect.

Are you getting different behavior or am I misunderstanding your case? Could you please provide a minimal example?

**Comment by elenigen** [comment]
Actually, I already tried to provide an example by applying only few  changes to your sample project, so just to make it clear, I provide again the code + request + actual response + expected response:
### Code
https://github.com/elenigen/graphql-spqr-samples

### Request
```js
mutation {
  createVendor(
    vendor: {
      id: 1
      name: "hello"
      address: {
        city: "mtl"
        postalCode:"h2c1f1"
        streetAndNumber: "123 asd"
      }
      productsInStock:[
        {
          stockSize: 1
          product:{
            id: 1
            name: "aaa"
            description:"asdas"
          }
        }
      ]
    }) {
    id
    name
  }
}
```

### Actual response
```js
{
  "data": null,
  "errors": [
    {
      "locations": null,
      "message": "invalid data",
      "errorType": null,
      "path": null,
      "extensions": null
    }
  ],
  "extensions": null
}
```

### Expected response
```js
{
  "data": {
    "createVendor": {
      "id": 1,
      "name": "hello"
    }
  },
  "errors": [
    {
      "locations": null,
      "message": "invalid data",
      "errorType": null,
      "path": null,
      "extensions": null
    }
  ],
  "extensions": null
}
```

I get the same situation if I use anonymous error class like you did, so let me know if you need more details.
Thanks again!

**Comment by kaqqao** [comment]
Oh, hahahaa, I see the problem now and it's absolutely silly :laughing:
The result is exactly as you expect it, but the `sanitize` method called after the execution by the `GraphQLSampleController` destroys the data, because it creates a new `ExecutionResult` with the error only. This is because it's just a quick hack from the time before `ExecutionResult#toSpecification` used to hide the stacktrace (the comment also says something to that effect). No reason to still exist...
I've fixed up [the demo code](https://github.com/leangen/graphql-spqr-samples/blob/master/src/main/java/io/leangen/spqr/samples/demo/controller/GraphQLSampleController.java#L65) just now to follow the times. It now just calls `toSpecification` and `sanitize` is removed.

Sorry for the confusion... The sample project always lags behind because we only really update it properly before public talks. But I'll make sure to update it more regularly from now on.
We're about to release a SPQR-powered Spring Boot starter, so we'll have samples showcasing that too.

**Comment by elenigen** [comment]
Thanks for the investigation, I know it's not an easy task to keep a sample project in sync with the original one.  One last point, I will try to debug this but when I apply the changes your suggested, it worked to send data + error, but if I add two errors in the list, for some reason, only the first one show up in the response, but it might be a graphql-java problem:

### Code
```java
    @GraphQLMutation(name = "createVendor")
    public DataFetcherResult<Vendor> createVendor(@GraphQLArgument(name = "vendor") Vendor vendor){
        Vendor createdVendor =  new Vendor((long) mockVendorStorage.size() + 1,
                vendor.getName(),
                vendor.getAddress());
        mockVendorStorage.add(createdVendor);
        List<GraphQLError> errors = new ArrayList<>();
        errors.add(new GenericGraphQLError("invalid data1"));
        errors.add(new GenericGraphQLError("invalid data2"));
        return new DataFetcherResult<>(createdVendor, errors);
    }
```
### Response
```
{
  "data": {
    "createVendor": {
      "id": 1,
      "name": "hello"
    }
  },
  "errors": [
    {
      "message": "invalid data1"
    }
  ]
}
```

**Comment by elenigen** [comment]
I was right, the source of the problem is from graphql-java and it was already fixed ([#888](https://github.com/graphql-java/graphql-java/issues/888)), so it would be part of the next release.

As reference, here's what I did to make it work:

### Code
```java
    @GraphQLMutation(name = "createVendor")
    public DataFetcherResult<Vendor> createVendor(@GraphQLArgument(name = "vendor") Vendor vendor){
        Vendor createdVendor =  new Vendor((long) mockVendorStorage.size() + 1,
                vendor.getName(),
                vendor.getAddress());
        mockVendorStorage.add(createdVendor);
        List<GraphQLError> errors = new ArrayList<>();
        errors.add(generateFakeError("invalid data1", "path1"));
        errors.add(generateFakeError("invalid data2", "path1"));
        return new DataFetcherResult<>(createdVendor, errors);
    }

    private GraphQLError generateFakeError(String message, String path) {
        return new GraphQLError() {
            @Override
            public String getMessage() {
                return "Something went wrong";
            }

            @Override
            public List<Object> getPath() {
                return Collections.singletonList(path);
            }

            @Override
            public List<SourceLocation> getLocations() {
                return null;
            }

            @Override
            public ErrorType getErrorType() {
                return ErrorType.ValidationError;
            }
        };
    }
```

### Request
```js
mutation {
  createVendor(
    vendor: {
      id: 1
      name: "hello"
      address: {
        city: "mtl"
        postalCode:"h2c1f1"
        streetAndNumber: "123 asd"
      }
      productsInStock:[
        {
          stockSize: 1
          product:{
            id: 1
            name: "aaa"
            description:"asdas"
          }
        }
      ]
    }) {
    id
    name
  }
}
```

### Response
```js
{
  "data": {
    "createVendor": {
      "id": 1,
      "name": "hello"
    }
  },
  "errors": [
    {
      "message": "invalid data1",
      "path": [
        "createVendor",
        "path1"
      ]
    },
    {
      "message": "invalid data2",
      "path": [
        "createVendor",
        "path2"
      ]
    }
  ]
}
```

Finally, I was able to make it work, thanks a lot for your help!
😃 🍻  🌴

**Comment by kaqqao** [comment]
Glad we got it sorted out! Cheers! :beer:
